[WebCore] Do not add the same StyleSheetContents repeatedly

Constellation opened the pull request:
#### 9010531b7cdcfd824b5429729e1047a41e05d93f
<pre>
[WebCore] Do not add the same StyleSheetContents repeatedly
<a href="https://bugs.webkit.org/show_bug.cgi?id=273455">https://bugs.webkit.org/show_bug.cgi?id=273455</a>
<a href="https://rdar.apple.com/127263420">rdar://127263420</a>

Reviewed by Antti Koivisto.

This is derived from [1]. In some cases, we are repeatedly having &lt;style&gt; with exact same content.
But fortunately, we are already deduping them via StyleSheetContents. This patch avoids adding the
same content &lt;style&gt; repeatedly in RuleSetBuilder.
Because this is very pathological pattern and putting a bit more cost on StyleScopeRuleSets etc.
caused rather regression, this patch just checks the one previous &lt;style&gt; and only handles the cases
where mediaQueries is empty. This one-previous-style-sheet pattern is also taken in [1].

[1]: <a href="https://chromium-review.googlesource.com/c/chromium/src/+/5385787">https://chromium-review.googlesource.com/c/chromium/src/+/5385787</a>

* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::appendAuthorStyleSheets):

Canonical link: <a href="https://commits.webkit.org/278184@main">https://commits.webkit.org/278184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c9a351c1d199dea1650a940a7d478600e728d44d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49755 "Failed to checkout and rebase branch from PR 27909") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29042 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52805 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52059 "Failed to checkout and rebase branch from PR 27909") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/35068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26659 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51854 "Failed to checkout and rebase branch from PR 27909") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/35068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/52805 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/35068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/52805 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8126 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/35068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/52805 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54578 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/24842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/52805 "Build is in progress. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/26956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7162 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleScopeRuleSets.cpp:
```diff
@@ -238,10 +238,19 @@ void ScopeRuleSets::appendAuthorStyleSheets(const Vector<RefPtr<CSSStyleSheet>>&
+    RefPtr<CSSStyleSheet> previous;
     for (auto& cssSheet : styleSheets) {
         ASSERT(!cssSheet->disabled());
+        if (previous) {
+            if (&previous->contents() == &cssSheet->contents() && previous->mediaQueries().isEmpty() && cssSheet->mediaQueries().isEmpty()) {
+                inspectorCSSOMWrappers.collectFromStyleSheetIfNeeded(cssSheet.get());
+                continue;
+            }
+        }
```
This could use a comment explaining why only checking previous is a decent optimization

Constellation committed:
[WebCore] Do not add the same StyleSheetContents repeatedly
https://bugs.webkit.org/show_bug.cgi?id=273455
rdar://127263420

Reviewed by Antti Koivisto.

This is derived from [1]. In some cases, we are repeatedly having <style> with exact same content.
But fortunately, we are already deduping them via StyleSheetContents. This patch avoids adding the
same content <style> repeatedly in RuleSetBuilder.
Because this is very pathological pattern and putting a bit more cost on StyleScopeRuleSets etc.
caused rather regression, this patch just checks the one previous <style> and only handles the cases
where mediaQueries is empty. This one-previous-style-sheet pattern is also taken in [1].

[1]: https://chromium-review.googlesource.com/c/chromium/src/+/5385787

* Source/WebCore/style/StyleScopeRuleSets.cpp:
(WebCore::Style::ScopeRuleSets::appendAuthorStyleSheets):

Canonical link: https://commits.webkit.org/278184@main
